File: crates/hushd/src/siem/transforms/cef.rs
use crate::siem::types::{Outcome, ResourceType, SecurityEvent, SecuritySeverity};

/// Convert a `SecurityEvent` to ArcSight-style CEF.
///
/// This is a lightweight mapping intended for log forwarding (not a full CEF dictionary coverage).
pub fn to_cef(event: &SecurityEvent) -> String {
    let version = "0";
    let device_vendor = "Clawdstrike";
    let device_product = "hushd";
    let device_version = event.agent.version.as_str();
    let signature_id = format!("{:?}", event.event_type);
    let name = event.decision.guard.clone();
    let severity = cef_severity(&event.decision.severity);

    let mut ext = vec![
        kv("eventId", &event.event_id.to_string()),
        kv("sessionId", &event.session.id),
        kv(
            "outcome",
            match event.outcome {
                Outcome::Success => "success",
                Outcome::Failure => "failure",
                Outcome::Unknown => "unknown",
            },
        ),
        kv("action", &event.action),
        kv("guard", &event.decision.guard),
        kv("reason", &event.decision.reason),
    ];

    match event.resource.resource_type {
        ResourceType::File => {
            if let Some(path) = &event.resource.path {
                ext.push(kv("filePath", path));
            }
        }
        ResourceType::Network => {
            if let Some(host) = &event.resource.host {
                ext.push(kv("dst", host));
            }
            if let Some(port) = event.resource.port {
                ext.push(kv("dpt", &port.to_string()));
            }
        }
        ResourceType::Process => {
            ext.push(kv("process", &event.resource.name));
        }
        ResourceType::Tool => {
            ext.push(kv("tool", &event.resource.name));
        }
        ResourceType::Configuration => {}
    }

    format!(
        "CEF:{version}|{device_vendor}|{device_product}|{device_version}|{signature_id}|{name}|{severity}|{}",
        ext.join(" ")
    )
}

fn cef_severity(sev: &SecuritySeverity) -> u8 {
    // CEF severity is 0-10.
    match sev {
        SecuritySeverity::Info => 1,
        SecuritySeverity::Low => 3,
        SecuritySeverity::Medium => 5,
        SecuritySeverity::High => 8,
        SecuritySeverity::Critical => 10,
    }
}

fn kv(key: &str, value: &str) -> String {
    // CEF extensions are key=value; escape = and \.
    let v = value.replace('\\', "\\\\").replace('=', "\\=");
    format!("{key}={v}")
}

File: packages/sdk/hush-go/siem/transforms/cef.go
package transforms

import (
	"fmt"
	"strings"

	"github.com/backbay-labs/clawdstrike-go/siem"
)

// CEF severity mapping.
var severityToCEF = map[string]int{
	"info":     1,
	"warning":  5,
	"warn":     5,
	"error":    8,
	"low":      3,
	"medium":   5,
	"high":     7,
	"critical": 10,
}

// ToCEF converts a SecurityEvent to Common Event Format (ArcSight) string.
// Format: CEF:0|vendor|product|version|signatureId|name|severity|extensions
func ToCEF(event siem.SecurityEvent) string {
	severity := 0
	name := event.EventType
	signatureID := event.EventType

	if event.Decision != nil {
		if s, ok := severityToCEF[event.Decision.Severity]; ok {
			severity = s
		}
		name = event.Decision.Guard
	}

	extensions := []string{
		fmt.Sprintf("rt=%d", event.Timestamp.UnixMilli()),
		fmt.Sprintf("externalId=%s", cefEscape(event.EventID)),
		fmt.Sprintf("outcome=%s", event.Outcome),
	}

	if event.Agent.ID != "" {
		extensions = append(extensions, fmt.Sprintf("suid=%s", cefEscape(event.Agent.ID)))
	}
	if event.Agent.Name != "" {
		extensions = append(extensions, fmt.Sprintf("suser=%s", cefEscape(event.Agent.Name)))
	}

	if event.Resource != nil {
		switch event.Resource.Type {
		case "file":
			extensions = append(extensions, fmt.Sprintf("filePath=%s", cefEscape(event.Resource.Path)))
		case "network":
			extensions = append(extensions, fmt.Sprintf("dhost=%s", cefEscape(event.Resource.Host)))
		case "tool":
			extensions = append(extensions, fmt.Sprintf("cs1=%s cs1Label=tool", cefEscape(event.Resource.Tool)))
		}
	}

	if event.Decision != nil && event.Decision.Message != "" {
		extensions = append(extensions, fmt.Sprintf("msg=%s", cefEscape(event.Decision.Message)))
	}

	return fmt.Sprintf("CEF:0|Backbay|Clawdstrike|1.0|%s|%s|%d|%s",
		cefEscape(signatureID),
		cefEscape(name),
		severity,
		strings.Join(extensions, " "),
	)
}

// cefEscape escapes special characters in CEF field values.
func cefEscape(s string) string {
	s = strings.ReplaceAll(s, `\`, `\\`)
	s = strings.ReplaceAll(s, `|`, `\|`)
	s = strings.ReplaceAll(s, `=`, `\=`)
	s = strings.ReplaceAll(s, "\n", `\n`)
	s = strings.ReplaceAll(s, "\r", `\r`)
	return s
}
